### PR TITLE
Ampers: Alex

### PR DESCRIPTION
**Pull request by brownav**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | On whether to include a room class or not -- I decided the class would be too scant to be of much use, and instead kept rooms as an attribute of Hotel::Admin.
Describe a concept that you gained more clarity on as you worked on this assignment. | Testing before writing methods to help gain clarity on an approach before coding, pseudocoding, and being more robust with testing edge-cases and parameters. 
Describe a _nominal test_ that you wrote for this assignment. | Testing if errors were raised for out-of-range or constraint-breaking inputs where used. 
Describe an _edge case test_ that you wrote for this assignment. | Testing if errors were raised for nil or missing inputs. 
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I improved on this aspect as the project went further along, as the methods became more complicated to construct, I took more effort to write pseudocode first, then construct tests, then fill in the pseudocode with actual code. This approach actually proved really helpful despite taking some deliberate getting used to at first.

### CODE REVIEW

File: USA [hunk 1]
```diff
@@ -0,0 +1 @@
+WHERE IS ADA?
```
Comment (CheezItMan) [inline USA:1]: Here!

1215 4th Ave
Suite 1050
Seattle, WA 98161

File: lib/admin.rb [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    end
+
+    def free_rooms_for_dates(days)
+      days = validate_dates(days)
+
+      @rooms = room_list
+      @blocks = []
+
+      @reservations.each do |reservation|
+        overlap = reservation.dates & days
```
Comment (CheezItMan) [inline lib/admin.rb]: This won't work unless `days` has an exact date of a reservation.  So this requires your reservations to keep all the dates it's reserved.  

However doing it this way does make & work, clever.  

You also will need to consider rooms in blocks however.

File: specs/admin_spec.rb [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    end
+
+    it "removes a room from rooms list" do
+      admin = Hotel::Admin.new
+      res1 = Hotel::Reservation.new(['12 May', '13 May'])
+      free_rooms = admin.rooms.length
+
+      admin.book_room(res1)
+
+      admin.rooms.length.must_equal free_rooms - 1
```
Comment (CheezItMan) [inline specs/admin_spec.rb:37]: This test doesn't make sense, the number of rooms doesn't change when a room is booked, the number of free rooms in a date range would change.

File: lib/admin.rb [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        if overlap.length > 0
+          @rooms.pop
```
Comment (CheezItMan) [inline lib/admin.rb:86]: Removing the room from the list doesn't make sense, since when I book a room the hotel doesn't lose a room permanently.

File: lib/admin.rb [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        raise ArgumentError.new("Not a valid reservation instance #{reservation}")
+      elsif free_rooms_for_dates(reservation.dates).length == 0
+        raise ArgumentError.new("No available rooms for these dates #{reservation.dates}")
+      end
+
+      @reservations << reservation
+      book_room(reservation)
+    end
+
+    def create_block_of_rooms(reservation, num_rooms)
```
Comment (CheezItMan) [inline lib/admin.rb]: What about:
1.  Creating a block with a discounted room rate.
1.  Is a block a `Reservation`?  
1.  Is a room really reserved, if it's in the block, or is it just being held for later?

**Comment by CheezItMan** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Design** |
Demonstrated classes having a single responsibility |  For the most part yes, More work should be handed off to reservation.  Also you are using a hash for a block.  That could very well be a class in it's own right.  
Demonstrated loose coupling |  Check
Methods demonstrate a good use of encapsulation, inputs and outputs | Some issues here, see my notes.  
Wave 1 requirements |  You've got some issues with your logic, specifically you're removing rooms permanently if they are ever reserved, even for one day.  
Wave 2 requirements |  See my notes.
Wave 3 requirements |  You seem to be confusing a block with a reservation.  Instead of holding rooms for future reservations you're taking the Block as a bulk reservation.  
Summary | There's some clever work here specifically with how you see if a reservation overlaps with a date, but there's also a lot of problematic code.  Checkout our [reference implementation](https://github.com/Ada-Developers-Academy/textbook-curriculum/blob/master/02-intermediate-ruby/exercises/hotel-revisited.md)
